include/openmc/weight_windows.h

```diff
@@ -0,0 +1,99 @@
+#define OPENMC_WEIGHT_WINDOWS_H
+
+#include <cstdint>
+#include <unordered_map>
+
+#include <hdf5.h>
+#include <pugixml.hpp>
+
+#include "openmc/constants.h"
+#include "openmc/memory.h"
+#include "openmc/mesh.h"
+#include "openmc/particle.h"
+#include "openmc/vector.h"
+
+namespace openmc {
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+//! Apply weight windows to a particle
+//! \param[in] p  Particle to apply weight windows to
+void apply_weight_windows(Particle& p);
+
+//! Free memory associated with weight windows
+void free_memory_weight_windows();
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+class WeightWindows;
+
+namespace variance_reduction {
+
+extern std::unordered_map<int32_t, int32_t> ww_map;
+extern vector<unique_ptr<WeightWindows>> weight_windows;
+
+} // namespace variance_reduction
+
+//==============================================================================
+//! Individual weight window information
+//==============================================================================
+
+struct WeightWindow {
+  double lower_weight {-1}; // -1 indicates not valid state
+  double upper_weight {1};
+  double survival_weight {0.5};
+  double weight_cutoff {WEIGHT_CUTOFF};
+  int max_split {1};
+
+  //! Whether the weight window is in a valid state
+  operator bool() const { return lower_weight >= 0.0; }
+};
+
+//==============================================================================
+//! Weight window settings
+//==============================================================================
+
+class WeightWindows {
+public:
+  // Constructors
+  WeightWindows();
+  WeightWindows(pugi::xml_node node);
+
+  // Methods
+
+  //! Set the weight window ID
+  void set_id(int32_t id = -1);
+
+  // NOTE: This is unused for now but may be used in the future
+  //! Write weight window settings to an HDF5 file
+  //! \param[in] group  HDF5 group to write to
+  void to_hdf5(hid_t group) const;
+
+  //! Return a weight window at the specified index
+  //! \param[in] p  Particle to get weight window for
+  WeightWindow get_weight_window(const Particle& p) const;
+
+  // Accessors
+  int32_t id() const { return id_; }
+  const Mesh& mesh() const { return *model::meshes[mesh_idx_]; }
+
+private:
+  // Data members
+  int32_t id_;                     //!< Unique ID
+  ParticleType particle_type_;     //!< Particle type to apply weight windows to
+  vector<double> energy_bins_;     //!< Energy bins [eV]
+  vector<double> lower_ww_;        //!< Lower weight window bounds
+  vector<double> upper_ww_;        //!< Upper weight window bounds
+  double survival_ratio_ {3.0};    //!< Survival weight ratio
+  double weight_cutoff_ {1.0e-38}; //!< Weight cutoff
```

Comment:
We could just get rid of the new variable defined in `constants.h` if 1e-38 is going to be hard-coded here anyway. Or make it something like `constexpr double DEFAULT_WEIGHT_CUTOFF` defined only in the class `WeightWindows`.

Reply:
I took a slightly different direction with this. The default weight cutoff is used in both `WeightWindow` and `WeightWindows`, so it's easiest to keep that value outside of the class. At that point, I think it makes sense to keep a value like that with the rest of our hardcoded constants and opted to do so.

Let me know what you think.

Reply:
Sounds reasonable enough!